pkg/syncer/specsyncer.go

```diff
@@ -143,11 +175,44 @@ func (c *Controller) ensureDownstreamNamespaceExists(ctx context.Context, downst
 	return nil
 }
 
-func (c *Controller) applyToDownstream(ctx context.Context, gvr schema.GroupVersionResource, downstreamNamespace string, upstreamObj *unstructured.Unstructured) error {
+func (c *Controller) notifySyncerOwnership(ctx context.Context, gvr schema.GroupVersionResource, upstreamObj *unstructured.Unstructured) error {
+	name := upstreamObj.GetName()
+	namespace := upstreamObj.GetNamespace()
+
+	// If the advanced scheduling feature is enabled, add the Syncer Finalizer to the upstream object
+	if advancedSchedulingFeatureEnabled {
+		upstreamFinalizers := upstreamObj.GetFinalizers()
+		hasFinalizer := false
+		for _, finalizer := range upstreamFinalizers {
+			if finalizer == SyncerFinalizerName(c.pcluster) {
+				hasFinalizer = true
+			}
+		}
+		if !hasFinalizer {
+			upstreamFinalizers = append(upstreamFinalizers, SyncerFinalizerName(c.pcluster))
+			upstreamObj.SetFinalizers(upstreamFinalizers)
```

Comment:
this is bad style. Never mutate inputs if it is not super clear that the func is intended to do that. `notifySyncerOwnership` does not suggest side-effects, actually not even that is does some client call. `ensureSyncerFinalizer` would be much better.

Reply:
this is even much worse if I am not mistaken: the upstream object comes directly from an informer. So this func mutates cached data, leading to inconsistencies, and also to races.

Reply:
deepcopied the object, and changed the name.